File: BDH.Rhino.Web.FrontEnd/src/app/shared/models/Vector2d.ts
import { IsPoint2d } from "./IsPoint2d";
import { Point2d } from "./Point2d";


export class Vector2d implements IsPoint2d {
    constructor(public x: number, public y: number) {
    }


    static fromPoints(tip: IsPoint2d, tale: IsPoint2d) {
        return new Vector2d(tip.x - tale.x, tip.y - tale.y);
    }

    getLength() {
        return Math.sqrt(Math.pow(this.x, 2) + Math.pow(this.y, 2));
    }

    normalize() {
        var length = this.getLength();

        return new Vector2d(
            this.x / length,
            this.y / length
        );
    }

    setLength(length: number) {
        var normalized = this.normalize();

        return new Vector2d(
            normalized.x * length,
            normalized.y * length
        );
    }

    rotate(radians: number) {
        var point = new Point2d(this.x, this.y);
        var anchor = new Point2d(0, 0);
        var rotated = point.rotateAround(anchor, radians);
        return new Vector2d(rotated.x, rotated.y);
    }
}

File: BDH.Shared.Domain.Geometry.Extensions/Vector2D.cs
namespace BDH.Shared.Domain.Geometry.Extensions
{
    public class Vector2D
    {
        private readonly Point2D point;

        public double X => point.X;
        public double Y => point.Y;
        public double Length => point.Length;


        public Vector2D(Point2D point)
        {
            this.point = point;
        }

        public Vector2D(double x, double y)
        {
            this.point = new Point2D(x, y);
        }



        public double CrossProduct(Vector2D other)
        {
            return (X * other.Y) - (Y * other.X);
        }
        public double DotProduct(Vector2D other)
        {
            return (X * other.X) + (Y * other.Y);
        }


        public Vector2D Reverse()
        {
            return new Vector2D(new Point2D(-X, -Y));
        }
        public Vector2D Normalize()
        {
            return new Vector2D(new Point2D(X / Length, Y / Length));
        }
        public Vector2D Multiply(double length)
        {
            return new Vector2D(new Point2D(X * length, Y * length));
        }
        public Vector2D Rotate(double angle)
        {
            var point = new Point2D(X, Y);
            var anchor = new Point2D(0, 0);
            var rotated = point.RotateAround(anchor, angle);
            return new Vector2D(rotated);
        }
        public double AngleTo(Vector2D vector2)
        {
            double sin = X * vector2.Y - vector2.X * Y;
            double cos = X * vector2.X + Y * vector2.Y;

            return Math.Atan2(sin, cos) * (180 / Math.PI);
        }
    }
}